### PR TITLE
Add `cupy.trim_zeros`

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**

### CODE REVIEW

**Comment by Dahlia-Chehata** [comment]
Testing on array of zeros
| n | numpy | cupy |
| :---: | :---: | :---: |
| 100 | 40.993 | 232.495 |
| 1000 | 390.164 | 235.997 |
| 10000 |  3950.250 | 236.753 |
| 100000 | 39445.247 | 645.311 |
| 1000000 | 384831.584 | 5876.618 |

Array with no leading or trailing zeros
| n | numpy | cupy |
| :---: | :---: | :---: |
| 100 | 2.769 | 279.485 |
| 1000 | 6.902 | 342.328  |
| 10000 | 58.066  | 865.582 |
| 100000 | 842.294 | 7035.852 |
| 1000000 | 12975.920 | 68595.949  |

The performance is worse than numpy when trailing or leading zeros represent a small fraction of the input size.

File: cupy/manipulation/add_remove.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import cupy
-
+from cupy.core._reduction import ReductionKernel
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:2]: Importing non-module symbols is prohibited according to the [coding guideline](https://docs-cupy.chainer.org/en/stable/contribution.html#coding-guidelines).
How about using `cupy.ReductionKernel` simply?

File: cupy/manipulation/add_remove.py [hunk 2]
```diff
@@ -13,7 +13,63 @@
 # TODO(okuta): Implement resize
 
 
-# TODO(okuta): Implement trim_zeros
+_first_non_zero_krnl = ReductionKernel(
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:16]: I feel `_first_nonzero_kernel` is a better name for this kernel.

File: cupy/manipulation/add_remove.py [hunk 3]
```diff
@@ -13,7 +13,63 @@
...
+    'raw S data, T indices, T len, int64 mode',
+    'T y',
+    '''mode? data[indices] ? indices : len
+           : data[len - 1 - indices] ? len - 1 - indices : T(-1)''',
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:20]: How about supporting only for `mode=0`, and calling `len - _first_nonzero_kernel(filt[::-1], ...) - 1` to get the value of `end`?

File: cupy/manipulation/add_remove.py [hunk 4]
```diff
@@ -13,7 +13,63 @@
...
+    'mode? min(a,b): max(a, b) ',
+    'y = a',
+    'mode ? len : -1',
+    'first_non_zero'
+)
+_first_non_zero_cplx_krnl = ReductionKernel(
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:26]: Could you explain why this kernel is needed? Why not use `_first_non_zero_krnl` also for complex inputs?

File: cupy/manipulation/add_remove.py [hunk 5]
```diff
@@ -13,7 +13,63 @@
 # TODO(okuta): Implement resize
 
 
-# TODO(okuta): Implement trim_zeros
+_first_non_zero_krnl = ReductionKernel(
+    'raw S data, T indices, T len, int64 mode',
+    'T y',
+    '''mode? data[indices] ? indices : len
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:19]: `ReductionKernel` unofficially supports `_j` as the index of the inputs. How about using it?

File: cupy/manipulation/add_remove.py [hunk 6]
```diff
@@ -13,7 +13,63 @@
...
+            'fb' default option trims the array from both sides.
+            'f' option trim zeros from front.
+            'b' option trim zeros from front
+
+    Returns:
+        cupy.ndarray: trimmed input
+
+    .. seealso:: :func:`numpy.trim_zeros`
+    """
+    filt = cupy.asarray(filt)
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:56]: We don't want to support non-cupy-ndarray inputs to avoid unintended host-device synchronization.

File: cupy/manipulation/add_remove.py [hunk 7]
```diff
@@ -13,7 +13,63 @@
...
+    start, end = cupy.asarray([0, filt.size - 1], dtype=int)
+    indices = cupy.arange(filt.size)
+    trim = trim.upper()
+    if 'F' in trim:
+        if filt.dtype.kind == 'c':
+            _first_non_zero_cplx_krnl(
+                filt.real, filt.imag, indices, filt.size, 1, start)
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:63]: How about writing like as follows simply?
```suggestion
            start = _first_non_zero_cplx_krnl(
                filt.real, filt.imag, indices, filt.size, 1)
```

File: tests/cupy_tests/manipulation_tests/test_add_remove.py [hunk 8]
```diff
@@ -29,3 +29,97 @@ def test_unique_inverse(self, xp, dtype):
     def test_unique_counts(self, xp, dtype):
         a = testing.shaped_random((100, 100), xp, dtype)
         return xp.unique(a, return_counts=True)[1]
+
+
+@testing.gpu
+class TestTrim_zeros(unittest.TestCase):
```
Comment (asi1024) [inline tests/cupy_tests/manipulation_tests/test_add_remove.py:35]: How about using `testing.parameterize` to parameterize the input arrays?

Comment (asi1024) [reply tests/cupy_tests/manipulation_tests/test_add_remove.py:35]: Please add a test for zero-size input and non-1dim input.

**Comment by asi1024** [comment]
Could you show us the benchmark script?

File: cupy/manipulation/add_remove.py (hunk 3 above)
Comment (Dahlia-Chehata) [reply cupy/manipulation/add_remove.py:20]: I thought filt[::-1] will launch kernels multiple times and that is why I changed it to indexing inside the kernel.

File: cupy/manipulation/add_remove.py (hunk 6 above)
Comment (Dahlia-Chehata) [reply cupy/manipulation/add_remove.py:56]: So list/sequence are not allowed  ? like trim_zeros([1,2,3])

File: cupy/manipulation/add_remove.py (hunk 4 above)
Comment (Dahlia-Chehata) [reply cupy/manipulation/add_remove.py:26]: It raises error for complex inputs

Comment (Dahlia-Chehata) [reply cupy/manipulation/add_remove.py:26]: I can use  _first_non_zero_cplx_krnl for both complex and non-complex inputs but I needed extra memory (cupy.zeros) for non-complex inputs to pass it as imag (as in the 1st commit)

**Comment by Dahlia-Chehata** [comment]
> Could you show us the benchmark script?

```
def compare(n):
    def generator(n, xp):
        x = xp.asarray([random.randint(1000) for x in range(1, n)])
        y = xp.zeros(shape=(n,))
        return x, y
    
     x1, y1 = generator(n,np)
     x2, y2 = generator(n,cp)
      
    np_ret = repeat(np.trim_zeros, (x1, 'fb'), n=10, n_warmup=10)
    cp_ret = repeat(cp.trim_zeros, (x2, 'fb'), n=10, n_warmup=10)
    print("NumPy non-zeros: ", np_ret.to_str(show_gpu=True))
    print("CuPy non- zeros:", cp_ret.to_str(show_gpu=True))

    np_ret0 = repeat(np.trim_zeros, (y1, 'fb'), n=10, n_warmup=10)
    cp_ret0 = repeat(cp.trim_zeros, (y2, 'fb'), n=10, n_warmup=10)
    print("NumPy zeros: ", np_ret0.to_str(show_gpu=True))
    print("CuPy zeros:", cp_ret0.to_str(show_gpu=True))
```

File: tests/cupy_tests/manipulation_tests/test_add_remove.py (hunk 8 above)
Comment (Dahlia-Chehata) [reply tests/cupy_tests/manipulation_tests/test_add_remove.py:35]: ndim input is not handled in numpy. Should I raise some error, flatten, or leave without handling ? The current version doesn't change non-1dim input.

File: cupy/manipulation/add_remove.py [hunk 9]
```diff
@@ -12,8 +13,62 @@
 
 # TODO(okuta): Implement resize
 
+_first_nonzero_krnl = core.ReductionKernel(
+    'raw S data, T indices, T len',
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:17]: `indices` is not needed anymore. `len` is always of int64 type.
```suggestion
    'raw S data, int64 len',
```

File: cupy/manipulation/add_remove.py [hunk 10]
```diff
@@ -12,8 +13,62 @@
...
+    'T y',
+    'data[_j] ? _j : len',
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:19]: How about comparing with zero explicitly? It will work also for complex inputs.
```suggestion
    'data[_j] == S(0) ? _j : len',
```

Comment (asi1024) [reply cupy/manipulation/add_remove.py:19]: Anyway, I feel `T` is preferred for the type name.

File: cupy/manipulation/add_remove.py [hunk 11]
```diff
@@ -12,8 +13,62 @@
...
+            'b' option trim zeros from front
+
+    Returns:
+        cupy.ndarray: trimmed input
+
+    .. seealso:: :func:`numpy.trim_zeros`
+    """
+    if filt.ndim > 1:
+        raise ValueError("Multi-dimensional trim is not supported")
+    start, end = cupy.asarray([0, filt.size - 1], dtype=int)
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:55]: This line is not needed.
```suggestion
```

File: cupy/manipulation/add_remove.py [hunk 12]
```diff
@@ -12,8 +13,62 @@
...
+        else:
+            start = _first_nonzero_krnl(filt, indices, filt.size)
+    if 'B' in trim:
+        if filt.dtype.kind == 'c':
+            end = filt.size - _first_nonzero_cplx_krnl(
+                filt[::-1].real, filt[::-1].imag, indices, filt.size) - 1
+        else:
+            end = filt.size - _first_nonzero_krnl(
+                filt[::-1], indices, filt.size) - 1
+    return filt[start:end + 1]
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:71]: Please remove redundant `-1` and `+1`s.
```suggestion
    return filt[start:end]
```

File: tests/cupy_tests/manipulation_tests/test_add_remove.py (hunk 8 above)
Comment (asi1024) [reply tests/cupy_tests/manipulation_tests/test_add_remove.py:41]: Please check both NumPy and CuPy raise the error of the same type.

File: cupy/manipulation/add_remove.py [hunk 13]
```diff
@@ -12,8 +13,62 @@
...
+
+
+def trim_zeros(filt, trim='fb'):
+    """Trim the leading and/or trailing zeros from a 1-D array or sequence.
+
+    Returns the trimmed array
 
-# TODO(okuta): Implement trim_zeros
+    Args:
+        filt(1D array or sequence): Input array
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:42]:
```suggestion
        filt (cupy.ndarray): Input array
```

File: cupy/manipulation/add_remove.py [hunk 14]
```diff
@@ -12,8 +13,62 @@
...
+        trim(str, optional):
+            'fb' default option trims the array from both sides.
+            'f' option trim zeros from front.
+            'b' option trim zeros from front
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:46]:
```suggestion
            'b' option trim zeros from back.
```

File: tests/cupy_tests/manipulation_tests/test_add_remove.py (hunk 8 above)
Comment (asi1024) [reply tests/cupy_tests/manipulation_tests/test_add_remove.py:41]: Please add a test for 0-dim input.

File: cupy/manipulation/add_remove.py [hunk 15]
```diff
@@ -12,8 +13,62 @@
 
 # TODO(okuta): Implement resize
 
+_first_nonzero_krnl = core.ReductionKernel(
+    'raw S data, T indices, T len',
+    'T y',
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:18]:
```suggestion
    'int64 y',
```

File: cupy/manipulation/add_remove.py [hunk 16]
```diff
@@ -12,8 +13,47 @@
...
+            'b' option trim zeros from back.
+
+    Returns:
+        cupy.ndarray: trimmed input
+
+    .. seealso:: :func:`numpy.trim_zeros`
+
+    """
+    if filt.ndim > 1:
+        raise ValueError("Multi-dimensional trim is not supported")
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:46]: We use single quotes conventionally.
```suggestion
        raise ValueError('Multi-dimensional trim is not supported')
```

**Comment by Dahlia-Chehata** [comment]
> Could you remeasure the performance numbers?

**non-zeros array**      
| n | numpy | cupy |              
| :---: | :---: | :---: |              
| 100 |  7.441 | 142.199 |     
| 1000 | 6.797 | 141.197 | 
| 10000 | 6.985 | 141.895 |  
| 100000 |  9.788  | 226.032  | 
| 1000000 | 7.401 | 2614.253 | 

 **zeros array**      
| n | numpy | cupy |              
| :---: | :---: | :---: |              
| 100 |  41.617 | 140.243 |     
| 1000 | 425.337 | 140.158 | 
| 10000 |  3902.633 | 139.555 |  
| 100000 | 40540.193 | 226.290 | 
| 1000000 | 386880.186 | 2657.208 |

File: cupy/manipulation/add_remove.py [hunk 17]
```diff
@@ -12,8 +13,47 @@
...
+        raise ValueError('Multi-dimensional trim is not supported')
+    if not filt.ndim:
+        raise TypeError('0-d array cannot be trimmed')
+    start = 0
+    end = filt.size
+    trim = trim.upper()
+    if 'F' in trim:
+        start = _first_nonzero_krnl(filt, filt.size)
+    if 'B' in trim:
+        end = filt.size - _first_nonzero_krnl(filt[::-1], filt.size)
```
Comment (asi1024) [inline cupy/manipulation/add_remove.py:55]: Please convert 0-dim cupy.ndarray to scalar before 1-size GPU operation for the performance?
```suggestion
        end = filt.size - _first_nonzero_krnl(filt[::-1], filt.size).item()
```
`start = _first_nonzero_krnl(filt, filt.size).item()` is also preferred for the consistency.

File: tests/cupy_tests/manipulation_tests/test_add_remove.py [hunk 18]
```diff
@@ -29,3 +32,58 @@ def test_unique_inverse(self, xp, dtype):
...
+    @testing.numpy_cupy_array_equal()
+    def test_trim_empty(self, xp, dtype):
+        a = testing.empty(xp, dtype)
+        return xp.trim_zeros(a, trim=self.trim)
+
+    @testing.for_all_dtypes()
+    def test_trim_zero_dim(self, dtype):
+        a = testing.shaped_arange((), cupy, dtype)
+        with pytest.raises(TypeError):
+            cupy.trim_zeros(a, trim=self.trim)
```
Comment (asi1024) [inline tests/cupy_tests/manipulation_tests/test_add_remove.py:83]: Please check both NumPy and CuPy raises the error of the same type like as follows:
```py
for xp in (numpy, cupy):
    a = testing.shaped_arange(...)
    with pytest.raises(TypeError):
        ....
```

File: tests/cupy_tests/manipulation_tests/test_add_remove.py [hunk 19]
```diff
@@ -29,3 +32,58 @@ def test_unique_inverse(self, xp, dtype):
...
+    @testing.numpy_cupy_array_equal()
+    def test_trim_back_zeros(self, xp, dtype):
+        a = xp.array([1, 0, 2, 3, 0, 5, 0, 0, 0], dtype=dtype)
+        return xp.trim_zeros(a, trim=self.trim)
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_equal()
+    def test_trim_empty(self, xp, dtype):
+        a = testing.empty(xp, dtype)
+        return xp.trim_zeros(a, trim=self.trim)
```
Comment (asi1024) [inline tests/cupy_tests/manipulation_tests/test_add_remove.py:77]: What is the role of this test?

Comment (Dahlia-Chehata) [reply tests/cupy_tests/manipulation_tests/test_add_remove.py:77]: I added it before the 0dim test was requested . Will eliminate it

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
The performance numbers show us that `cupy.trim_zeros` of this PR performs well for large trimming size :smile:

**Comment by asi1024** [comment]
LGTM. Thank you for your contribution!
